### ISSUE TITLE
Add support for the playback sync

### ISSUE THREAD

**Issue by Razzeee** [Razzeee/script.trakt#93]
Add settings and sync for playback progress. If it's possible to get these imported to kodi.

http://docs.trakt.apiary.io/#reference/sync/playback/get-playback-progress

**Comment by freeman212** [comment]
I would love this :+1:

**Comment by s7eele** [comment]
👍+1

**Comment by Razzeee** [comment]
http://forum.kodi.tv/showthread.php?tid=161912

**Comment by Razzeee** [comment]
I'm giving this a little thought already, so which would you prefere? Overwriting playback progress in kodi or leaving it as if, even if trakt got one set?

**Comment by freeman212** [comment]
overwriting playback progress in Kodi. could be that you watched a bit further on a different device in the meantime.

**Comment by Razzeee** [comment]
Yes, thats what I'm thinking, it should always be better to get the trakt one

**Comment by freeman212** [comment]
but that's on the to-do list after the current version is submitted to the Kodi repo, right?

**Comment by Razzeee** [comment]
@freeman212 
laying some groundwork: https://github.com/fuzeman/trakt.py/issues/16
and have created a seperate branch for that (it will likely be broken)
but I'm running out of things I can do to the dev.

**Comment by freeman212** [comment]
good news, a separate branch seems like the best option right now, because the dev branch is considered almost stable right now as you're preparing for the release ;-)

**Comment by Razzeee** [comment]
So playback sync for movies should be working. Very rough, I need to redo the percentages and dialogtexts of the progressdialog.
Not really sure how accurate the progress sync, haven't had time to check it. Feedback would be nice, don't know when I can get into it again.

**Comment by s7eele** [comment]
testing now, will provide feedback as appropriate

**Comment by Razzeee** [comment]
I suspect the sync to be quiet off. Need a solution for this: http://forum.kodi.tv/showthread.php?tid=218144&pid=1936058#pid1936058

**Comment by s7eele** [comment]
@Razzeee wrt movie playback sync it seemed really accurate except for a couple things ive noted before, specifically that the progress is not updated on trakt website when using either the progress bar or +/-  30 secs key to go back and forth, it takes a refresh on the website to read the new location after using either method above, progress continues on trakt website but its off until a forced refresh

i hope i havent completely misunderstood what you wanted feedback on, this seemed to be working with last -dev branch i was using as well, i realize this issue is bigger than movie playback progress so really hoping i didnt waste everyones time (the movie i tested was 'brick mansions' which had no redeeming value whatsoever, painful waste of my time at least lol)

**Comment by Razzeee** [comment]
Can't do anything about the trakt website ;) 
The case we need to test is basically, watch something with scrobble on. Skip somewhere, wait a bit and write down the minutes and seconds. Exit. 
On another device or installation, do a sync with playback enabled and check if the movie now has a resume point. Check how much it differs from the one you've written down.

**Comment by s7eele** [comment]
roger that, im working on testing that scenario now but it will be tomorrow before i can post any results, right now manual sync is hanging at 99% on my windows box so i can only work with my android boxes, will look into the windows issue and let you know if its something i need help with, ill update with the actual feedback you want soon :-)

**Comment by s7eele** [comment]
a lot more testing to do but i verified that a resume point was set on device 1 (windows) and subsequently detected by device 2 (android), resume point was set at 1:31:51 on device 1 while device 2 detected the resume point at 1:31:34 (one brief preliminary test, some issues seen, will finish tomorrow but thought it was worth reporting that a resume point was set and detected with only 17 second variance)

**Comment by Razzeee** [comment]
@s7eele 17 seconds would be fine for me I think.
would be nice if the episodes were also working, but there seems to be something wrong with the data from kodi

**Comment by freeman212** [comment]
a little bit of overlapping is really useful imo so you get back to the story.

**Comment by Razzeee** [comment]
@freeman212 
yep, think so too, we'll have to test some more, just to see if this is consistant.

**Comment by Razzeee** [comment]
Would be greate if some native english speakers could check this commit and add some comments.
Not really sure about the translation of "playback progress" etc
https://github.com/Razzeee/script.trakt/commit/989bbb04bdf20e0bfbfa05c7ad38b163b1e009e4

**Comment by s7eele** [comment]
roger that

**Comment by Razzeee** [comment]
Seems like there is no way to get this working for episodes...
Check this: http://trac.kodi.tv/ticket/14524

**Comment by Razzeee** [comment]
So if the kodi team isn't doing anything about the episode runtime I will deactivate that option, test movies a little bit and merge it to the dev branch

**Comment by s7eele** [comment]
understood @razzeee its probably going to take me the rest of today to get really firm results, just an fyi

**Comment by s7eele** [comment]
@razzeee the good news is that playback progress points for movies (resume points) are being set and read properly (max variance of about -60 seconds, always in the right direction so far, providing the overlap discussed above) but ive encountered other issues that require a little more research before i can say whats causing them, could be something on my end

in one case i stopped playback on device 1 and the resume point was seen on device 2 but when selecting resume vs. start from beginning it started from beginning anyway, in another case it worked the same wrt resume points but after selecting resume this time it just hung and i was forced to kill kodi

@tknorris implemented this same (or very similar) functionality in SALTS, i believe he uses trakt bookmarks in his situation but im thinking he may have worthwhile input, hence the notification, apologies in advance if im wrong

i will continue testing and update as appropriate

**Comment by tknorris** [comment]
The only thing SALTS really does is read and use trakt bookmarks, it makes no attempt to set them. It assumes that the official addon will set them as needed when playback stops. In fact, if the official addon starts setting internal xbmc bookmarks I'm probably going to have to make changes to the logic to make everything work smoothly in SALTS.

I did a lot of work w/ XBMC bookmarks for 1Channel, and after crawling through the XBMC code, my conclusion was there was no way to make XBMC resume a strm file at the correct resume point using it's internal bookmarks. You can see the results of my research in this issue: https://github.com/bstrdsmkr/1Channel/issues/107#issuecomment-44769311 The only thing I could ever get to work was custom bookmarks.

Note, razzee, you should be able to get the runtime on and video using the method that SALTS does here: https://github.com/tknorris/salts/blob/master/service.py#L120 It basically just monitors the playback and then merges the playback position at stop time with data that was passed to the window.

**Comment by tknorris** [comment]
Reading back through the comments on this issue, it also looks like there is some question about how to do the actual resume. What SALTS does is gets the bookmark % from trakt during starting a stream, then once playback actually starts, does a seek based on the actual length of the playing video and the progress % the trakt bookmark returned. That logic is here: https://github.com/tknorris/salts/blob/master/service.py#L83 and just before that is how I get the total length of the playing video.

I'm not sure if any of that helps or not!

**Comment by Razzeee** [comment]
It probably doesn't as we aren't playing the video in the moment we want to set the resume point. 
As we get all resume points from trakt, it is possible there are hundreds of them, which all should be set by the end of the sync.

**Comment by s7eele** [comment]
awesome information for me to review and get smarter if nothing else but i have to believe some of it will apply, thanks again tknorris

just an fyi since .strm files came up, i am including both in my testing atm, let me know if  should just focus on local razzeee

**Comment by tknorris** [comment]
Ah, I see. So, you aren't trying to set the resume points of the current video, you are trying to sync all of xbmc's interal bookmarks with all of trakt's bookmarks and vice versa. I haven't even thought about how that would work....

**Comment by Razzeee** [comment]
Exactly. It seems to work fine for movies, but the episodes code on kodi seems to be broken, check the trac ticket for details. 

@s7eele
Just write down if it was on strm or locals, if you test. Both would be great, but I don't know if I can tackle problems with strm if they show up

**Comment by tknorris** [comment]
That trac ticket seems to say if you pull the streamdetails that you can get the durations. Have you tried that? streamdetails would probably be better anyway because each episode length actually varies while there is only one "duration" value for an entire series from tvdb.

**Comment by s7eele** [comment]
roger @razzeee, exactly what ive been doing, bbl

**Comment by Razzeee** [comment]
@tknorris my tests showed completely empty streamdetails. I just need one that works. On movies it's the runtime which seems to be the real length from file not scrapped. 
But maybe I did something wrong.

**Comment by tknorris** [comment]
Try playing the episode once and then check the results of the call. SALTS gets episode details and I've found that the stream details will always be filled in once the video has bee played once. https://github.com/tknorris/salts/blob/master/scrapers/local_scraper.py#L60

**Comment by s7eele** [comment]
let me know if there is anything specific you want me to do wrt episodes

**Comment by s7eele** [comment]
@razzeee since updating to your last commit i havent experienced any of the issues i was seeing before, so far it looks really good, ive had two playbacks resume properly in the device to device scenario, still within a 45 - 60 second variance (probably less) but now ive seen it go both ways (+ or -), not complaining im just following up on the previous 'overlap' discussion

i need to test a little more including the .strm file scenario, will post final results with detail on timing and testing scenarios when done, if all continues to go the way it is now i should be able to say 'good to go' tonight or early tomorrow, for what thats worth lol

**Comment by Razzeee** [comment]
I'm not really seeing a fix for strm files, if I understood @tknorris post correctly. So we will have to leave them out.

You were playing back the same file each time? One thats on the smb share? (just to be sure)

**Comment by s7eele** [comment]
understood on .strm, just being thorough

wrt smb, it was the same file in each case, the file is on local drive of device 1 (window/pc) which shares it via smb to device 2 (android/ouya), there were more than two succesfull tests i just didnt have the timing details for those so didnt list them (approx 5), note it WAS tested in both directions

i suppose i should place a physical copy of the same movie on each device and test that scenario, ill run that and a few more tests in the background this afternoon to be certain but movie plaback progress sync is looking good to me (w/o .strm support for now :-\)

p.s. - @razzeee let me know if there are other scenarios you want tested

**Comment by Razzeee** [comment]
You don't need to have the file local the important part was that is is the same file with the same lenght.

Whats bothering me is that it can be in the direction of the end, should probably figure that one out before we do anything else.

**Comment by s7eele** [comment]
roger that and indeed it was the same file, i verified that the total runtime detected was the same on both devices

yeah, resuming at a point later than when it ended could cause problems, ill look into it and gather some logs in case you want to look at one or more

p.s. - i did save logs as i was testing yesterday so if you want to look at any just let me know (im looking through the logs as i have time, if i see anything that looks worthwhile ill let you know, im just using them for learning purposes atm (especially regarding .strm files, want to solidify my understanding)

**Comment by Razzeee** [comment]
Something really really weird happens with the percentage calculation here https://github.com/Razzeee/script.trakt/blob/playlist-sync/sync.py#L806
For some reason it gives me a different value when I try to calculate it with the windows calc...

**Comment by tknorris** [comment]
You are probably getting different results because that /100 makes the expression do integer division. If you replace it with 100.0 it should make the expression float division and give you the results you expect.

**Comment by Razzeee** [comment]
Thanks @tknorris not really sure how I missed that...
My first test are very promissing, showing a difference of 1 sec.

**Comment by Razzeee** [comment]
@s7eele 
another test that would be nice is a check if sync is still working correctly overall, with this branch.

**Comment by s7eele** [comment]
awesome, i was hopeful you/tk could offer some insight here but i was also hesitant to bug you/him with another notification

two great minds are better than one

standing by, havent looked at commits yet but will grab when able and do some more testing

**Comment by Razzeee** [comment]
The last second is probably due to the runtime on sync not having digits after the dot. Will probably have to round down at scrobble time to fix that

**Comment by s7eele** [comment]
ive been up way too long to try to write the details @razzeee but to summarize:

everything seems to be functional still, movie playback progress sync for locally stored movies worked consistently, however in 2 of 5 tests the resume point was after the pause/stop point set

as you are aware this means the user will miss a few seconds of playback, im in the middle of one last test and then plan to sleep for a bit, ill post detailed timings for each test later if still needed, i saw a max variance of 45-60 seconds but the norm was much lower, i dont have any resoning to back it up with yet but in my testing the problematic cases (resume later than pause) were all ones in which playback was started and stopped on device 2 (windows/local drive) and resumed on device 1 (android/smb), not sure if smb is causing a problem but i will be testing with a local copy of the same file on each device when i get a chance

i saw you mention rounding down the last second since its not available (think that was it) but i dont know  if any other alternatives are available to you that would help here, also not sure what acceptable tolerances would be, obviously you always want to resume playback before the stopping point but not sure if you would delay release or just make sure to manage expectations

bbl with more detail

**Comment by Razzeee** [comment]
You might want to pull the latest version first, the difference should be much smaller. 
I would like it to resume where I it was, we can add a setting for go back X seconds later.

**Comment by s7eele** [comment]
i was using commit 0817925, was that the wrong one @razzeee?

i just noticed another commit so will test later today after downloading (will be a couple hours before i can get back to it so no worries, ill make sure i download latest first, always look for most recent)

**Comment by Razzeee** [comment]
That one you referenced should have a difference of 1 sec, but maybe I'm not testing with the correct movies :/

The last commit brought us down to 0 sec difference.

Will do some more testing, especially if the other sync types remain unbroken.

**Comment by s7eele** [comment]
roger, ill run tests again shortly, shouldnt take as long this time

in your testing are u using a remote file system (nfs, smb, etc.) at all?

**Comment by Razzeee** [comment]
I'm always using my Nas via nfs. All my tests are ui only.

**Comment by s7eele** [comment]
roger, my nas died and i havent setup my new one yet, might make things more consistent (vs local drive on one device and smb share on the other)

**Comment by Razzeee** [comment]
Shouldn't make a difference, if really is the same file (same lenght)

**Comment by s7eele** [comment]
@razzeee give me an hour or two (maybe less) to run a couple more tests but the first test with latest version was spot on, bbiab

**Comment by s7eele** [comment]
@razzeee my friend i think youve nailed it wrt movie playback progress sync (local media only), i know thats not a surprise to you, just needed verification

everything is perfect, timing is accurate to the second every time but i am seeing about a 3-4 second delay in playback start on my ouya (playing media from an smb share), the context menu 'resume from' item and the 'resume from' pop-up messages are exactly correct but by the time i actually see playback start on my ouya i am generally 3-4 seconds past the resume point (happened on 3 of 4 tests), i realize this isnt within your control just wanted you to be aware of it as you look into setting a 'jumpback' so there is a little overlap, my pc (playing media from local drive) does not suffer from this 'lag', movies are resuming at the same exact second playback was ended

let me know if there is anything specific you need tested or tried out, ill continue testing overall functionality, thx again for everything

**Comment by Razzeee** [comment]
I'm pretty sure about the rest of the sync. Have done multiple test with this branch, as I try to find which movie in my collection is not beeing collected 601 collected on trakt vs 602 on kodi :/

**Comment by s7eele** [comment]
got it, i meant testing anywhere in the add-on but what i think im hearing is this is one of the last areas needing testing, i have a couple of abnormalities to chase down as well but likely to be something on my side, let me know if you find anything you want to point me at, peace

**Comment by Razzeee** [comment]
Got my problem, had a duplicate of one movie. So sync seems to be fine.
Will merge later

**Comment by s7eele** [comment]
awesome, your the man!

**Comment by s7eele** [comment]
@razzeee not sure you want to bother with it since ive only seen it this one time (could be an anommaly) but it was strange enough i wanted to mention it, an episode i was watching ended and i didnt see a rating dialogue pop-up, i noticed that a process kicked off right after playback ended so figured it was just lagging a bit, been several minutes now and never saw rating dialogue, looked on trakt to see if it showed up in history and it showed it was still playing @97% complete

**Comment by s7eele** [comment]
above issue is really weird but i havent seen it yet today so will wait on opening an issue (let me know if you want to look at logs from when it happened last night @razzeee), ive got some minor changes to the readme, should i go ahead and do a pull request or just pass them to you, dont want it to look like im just trying to get on contributor list, thx

**Comment by Razzeee** [comment]
Go ahead do a pull, that's a nice training for you ;) 
On the other issue the last two commits should really improve this case. 
But I assume it's still happening, can you try to force it to happen with the newest and provide a log?
Will try to do the same

**Comment by s7eele** [comment]
will do, i just pulled and installed latest, absolutely right about the training aspect, thx

**Comment by s7eele** [comment]
okay @razzeee, quick report without logs atm cause im in the middle of dealing with real life crap, vehicle issues

not sure im glad but i recreated it, today i had been watching different episodes of same show and no problems, played one of the episodes i watched last night without any interference (no pause, etc., staight playback) and same thing happened, playback ends in kodi but progress on trakt is at 78%, progress continues to increment until reaching 100% in trakt, i let kodi sit idle, im guessing it may act differently if i started something else while the false progress was still active on trakt website

final scrobble is never displayed in kodi so assuming it is never sent, rating dialogue is never received

so it looks like it could be episode specific, im also playing through genesis so that could be a factor, ill try running same episode through salts and asclibrary item to see if it makes a difference, ive also got the logs but i cant do anything until i take care of some vehicle related issues, shouldnt be too long and ill be available on my tablet

heres the tv show and episode numbers but dont judge me on my taste in shows, im running out (lol):

king of the nerds s01e07 and s01e08

as mentioned above, ive played two or three more episodes without issue but as soon as i went back to s01e07 it happened, ill be back shortly

**Comment by s7eele** [comment]
@razzeee details of whats happening and what ive found is available above, ive been able to reproduce it across two add-ons (genesis/salts), ive captured the log and identified the relevant portion so ill open an issue (unless i hear otherwise from you), it may turn out to be an issue outside of trakt (tvdb, etc) but i havent been able to determine that yet

it may be a while before i can actually open the issue or the pull request we discussed (btw, noticed one of the primary corrections was already fixed so wont be much of a change)

thx again for walking me through a lot of this

p.s. - going through the log i found the following entry that appeared within the relevant area, thought it might mean something to you:

> 18:23:23 T:1869480576 WARNING: static unsigned int CVideoInfoTag::GetDurationFromMinuteString(const string&) <runtime> should be in minutes. Interpreting 'None' as 0 minutes

**Comment by Razzeee** [comment]
Just a quick FYI. 
The website can't detect that the playback is over, without you refreshing it. 
It would be technically possible for the trakt guys to detect it, but it would be quite a waste. 

So be sure to refresh the page after playback stops.
